jetty-util/src/main/java/org/eclipse/jetty/util/thread/MonitoredQueuedThreadPool.java

```diff
@@ -0,0 +1,217 @@
+//  All rights reserved. This program and the accompanying materials
+//  are made available under the terms of the Eclipse Public License v1.0
+//  and Apache License v2.0 which accompanies this distribution.
+//
+//      The Eclipse Public License is available at
+//      http://www.eclipse.org/legal/epl-v10.html
+//
+//      The Apache License v2.0 is available at
+//      http://www.opensource.org/licenses/apache2.0.php
+//
+//  You may elect to redistribute this code under either of these licenses.
+//  ========================================================================
+//
+
+package org.eclipse.jetty.util.thread;
+
+import java.util.concurrent.TimeUnit;
+import java.util.concurrent.atomic.AtomicInteger;
+import java.util.concurrent.atomic.AtomicLong;
+import java.util.concurrent.atomic.LongAdder;
+
+import org.eclipse.jetty.util.Atomics;
+import org.eclipse.jetty.util.BlockingArrayQueue;
+import org.eclipse.jetty.util.annotation.ManagedAttribute;
+import org.eclipse.jetty.util.annotation.ManagedObject;
+import org.eclipse.jetty.util.annotation.ManagedOperation;
+
+@ManagedObject
+public class MonitoredQueuedThreadPool extends QueuedThreadPool
+{
+    private final LongAdder tasks = new LongAdder();
+    private final AtomicLong maxTaskLatency = new AtomicLong();
+    private final LongAdder totalTaskLatency = new LongAdder();
+    private final MonitoringBlockingArrayQueue queue;
+    private final AtomicLong maxQueueLatency = new AtomicLong();
+    private final LongAdder totalQueueLatency = new LongAdder();
+    private final AtomicInteger threads = new AtomicInteger();
+    private final AtomicInteger maxThreads = new AtomicInteger();
+
+    public MonitoredQueuedThreadPool()
+    {
+        this(256);
+    }
+
+    public MonitoredQueuedThreadPool(int maxThreads)
+    {
+        super(maxThreads, maxThreads, 24 * 3600 * 1000, new MonitoringBlockingArrayQueue(maxThreads, 256));
+        queue = (MonitoringBlockingArrayQueue)getQueue();
+        setStopTimeout(2000);
+    }
+
+    @Override
+    public void execute(final Runnable job)
+    {
+        final long begin = System.nanoTime();
+        super.execute(new Runnable()
+        {
+            @Override
+            public void run()
+            {
+                long queueLatency = System.nanoTime() - begin;
+                tasks.increment();
+                Atomics.updateMax(maxQueueLatency, queueLatency);
+                totalQueueLatency.add(queueLatency);
+                Atomics.updateMax(maxThreads, threads.incrementAndGet());
+                long start = System.nanoTime();
+                try
+                {
+                    job.run();
+                }
+                finally
+                {
+                    long taskLatency = System.nanoTime() - start;
+                    threads.decrementAndGet();
+                    Atomics.updateMax(maxTaskLatency, taskLatency);
+                    totalTaskLatency.add(taskLatency);
+                }
```

Comment:
I think it would also be interesting to set a thread local here, so we can measure how long after finishing one task before this thread picks up the next task.  If this is high it could indicate that some threads are being kept out from grabbing jobs by other busy threads... (I wonder if we are using fair locks?).   Anyway, set nanoTime into a thread local and check it on the next job.

Reply:
If the server is idle, the value reported will be large not because threads are kept out, but just because there are no tasks - so I am not sure we will have any value out it this value.

We don't use fair locks because they are way slower than non-fair locks.

Reply:
but if the server is busy and this value is still large, then than tells us something. But perhaps you are right and we don't need this until we suspect there is a fixable problem